mlflow/pyspark/ml/__init__.py

```diff
@@ -218,18 +346,79 @@ def _log_pretraining_metadata(estimator, params):
+                estimator_pipeline_hierarchy = _get_pipeline_stage_hierarchy(
+                    estimator.getEstimator())
+                try_mlflow_log(
+                    mlflow.log_dict,
+                    estimator_pipeline_hierarchy,
+                    artifact_file="estimator_pipeline_hierarchy.json"
+                )
+
         # Chunk model parameters to avoid hitting the log_batch API limit
         for chunk in _chunk_dict(
-            _get_instance_param_map(estimator), chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
+            param_map, chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
         ):
             truncated = _truncate_dict(chunk, MAX_ENTITY_KEY_LENGTH, MAX_PARAM_VAL_LENGTH)
             try_mlflow_log(mlflow.log_params, truncated)
 
         try_mlflow_log(mlflow.set_tags, _get_estimator_info_tags(estimator))
 
     def _log_posttraining_metadata(estimator, spark_model, params):
-        # TODO: Log nested runs for spark ml tuning estimators
-        #   (CrossValidator/TrainValidationSplit)
+
+        if _is_parameter_search_estimator(estimator):
+            try:
+                # Fetch environment-specific tags (e.g., user and source) to ensure that lineage
+                # information is consistent with the parent run
+                child_tags = context_registry.resolve_tags()
+                child_tags.update({MLFLOW_AUTOLOGGING: AUTOLOGGING_INTEGRATION_NAME})
+                _create_child_runs_for_parameter_search(
+                    parent_estimator=estimator,
+                    parent_model=spark_model,
+                    parent_run=mlflow.active_run(),
+                    child_tags=child_tags
+                )
+            except Exception as e:
+                import traceback
+                msg = (
+                    "Encountered exception during creation of child runs for parameter search."
+                    " Child runs may be missing. Exception: {}".format(traceback.format_exc())
+                )
+                _logger.warning(msg)
+
+            estimator_param_maps = _get_tuning_param_maps(estimator)
+
+            metric_key = estimator.getEvaluator().getMetricName()
+            if isinstance(spark_model, CrossValidatorModel):
+                metric_key = 'avg_' + metric_key
+                metrics = spark_model.avgMetrics
+            elif isinstance(spark_model, TrainValidationSplitModel):
+                metrics = spark_model.avgMetrics
+
+            search_result = []
+            for i in range(len(estimator_param_maps)):
+                search_result.append({
+                    'param_maps': estimator_param_maps[i],
+                    metric_key: metrics[i]
+                })
+            try_mlflow_log(
+                mlflow.log_dict,
+                search_result,
+                artifact_file="search_result.json"
+            )
+
         if log_models:
             if _should_log_model(spark_model):
```

Comment:
For sklearn, we log the CV model *and* the best model as separate artifacts. This allows users to more easily register the best model with the MLflow Model Registry. Can we do that here too?